plugins/azure/resoto_plugin_azure/resource/base.py

```diff
@@ -187,11 +216,13 @@ def collect(cls: Type[AzureResourceType], raw: List[Json], builder: GraphBuilder
             # map from api
             instance = cls.from_api(js)
             instance.pre_process(builder, js)
-            # add to graph
-            if (added := builder.add_node(instance, js)) is not None:
-                # post process
-                added.post_process(builder, js)
-                result.append(added)
+            # Check if there is a filter condition specified before adding the instance to the graph node.
+            if instance.should_add_to_node(builder, js, pre_items):
```

Comment:
I think that validate virtual machine sizes before we add to node them is better way. Because If we will iterate our nodes after collecting them, we would do more actions and spent more time.

Reply:
This only works if all virtual machines are collected beforehand.
I see you introduced `pre_collect`.  This might be a valid solution if we have one or two dependencies. But we actually have an arbitrary number of dependencies. Solving them by collecting them in different steps will make the code hard to understand and the collection time long (since only some things can be gathered in parallel). 

Fetching resources, adding them to the graph, and removing them in case they are not connected feels like unnecessary work. But it allows us to fetch everything in parallel and reason about the state that is collected. 

I would opt to keep the two steps we have:
- collect all resources
- process and connect